BUG: trying to downcast (Slice(UInt8) | Nil) (Crystal::MixedUnionType) <- Slice(UInt8) (Crystal::GenericClassInstanceType)

I just updated Lucky to Ameba 1.5.0, and when running my CI against Crystal nightly, I got this error.

https://github.com/luckyframework/lucky/actions/runs/5722039908/job/15504511981?pr=1818#step:4:37

```
Installing ameba (1.5.0)
Postinstall of ameba: shards build -Dpreview_mt
Failed postinstall of ameba on shards build -Dpreview_mt:
Dependencies are satisfied
Building: ameba
Error target ameba failed to compile:
BUG: trying to downcast (Slice(UInt8) | Nil) (Crystal::MixedUnionType) <- Slice(UInt8) (Crystal::GenericClassInstanceType) (Exception)
  from /crystal/src/compiler/crystal/codegen/cast.cr:521:5 in 'downcast_distinct'
  from /crystal/src/compiler/crystal/codegen/cast.cr:297:15 in 'downcast:extern'
  from /crystal/src/compiler/crystal/codegen/codegen.cr:1216:17 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/semantic/bindings.cr:16:7 in 'codegen_cond'
  from /crystal/src/compiler/crystal/codegen/codegen.cr:908:7 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/enumerable.cr:463:7 in '??'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/codegen.cr:2313:9 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/codegen.cr:2313:9 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/enumerable.cr:463:7 in '??'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/codegen.cr:2313:9 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/enumerable.cr:463:7 in '??'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/codegen.cr:2313:9 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/enumerable.cr:463:7 in '??'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/ast.cr:81:7 in 'codegen_fun'
  from /crystal/src/compiler/crystal/codegen/fun.cr:60:3 in 'target_def_fun'
  from /crystal/src/compiler/crystal/codegen/call.cr:437:12 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/semantic/bindings.cr:16:7 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/semantic/bindings.cr:16:7 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/enumerable.cr:463:7 in '??'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/ast.cr:81:7 in 'codegen_fun'
  from /crystal/src/compiler/crystal/codegen/fun.cr:60:3 in 'target_def_fun'
  from /crystal/src/compiler/crystal/codegen/call.cr:437:12 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/ast.cr:81:7 in 'codegen_fun'
  from /crystal/src/compiler/crystal/codegen/fun.cr:60:3 in 'target_def_fun'
  from /crystal/src/compiler/crystal/codegen/call.cr:437:12 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/codegen.cr:2330:7 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/codegen.cr:2313:9 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/enumerable.cr:463:7 in '??'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/exception.cr:92:7 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/enumerable.cr:463:7 in '??'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/codegen.cr:2313:9 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/enumerable.cr:463:7 in '??'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/ast.cr:81:7 in 'codegen_fun'
  from /crystal/src/compiler/crystal/codegen/fun.cr:60:3 in 'target_def_fun'
  from /crystal/src/compiler/crystal/codegen/call.cr:437:12 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/codegen.cr:2330:7 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/codegen.cr:2313:9 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/codegen.cr:2330:7 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/semantic/bindings.cr:16:7 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/enumerable.cr:463:7 in '??'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/ast.cr:81:7 in 'codegen_fun'
  from /crystal/src/compiler/crystal/codegen/fun.cr:60:3 in 'target_def_fun'
  from /crystal/src/compiler/crystal/codegen/call.cr:437:12 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/semantic/bindings.cr:16:7 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/enumerable.cr:463:7 in '??'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/ast.cr:81:7 in 'codegen_fun'
  from /crystal/src/compiler/crystal/codegen/fun.cr:60:3 in 'target_def_fun'
  from /crystal/src/compiler/crystal/codegen/call.cr:437:12 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/ast.cr:81:7 in 'codegen_fun'
  from /crystal/src/compiler/crystal/codegen/fun.cr:60:3 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/call.cr:316:5 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/ast.cr:81:7 in 'codegen_fun'
  from /crystal/src/compiler/crystal/codegen/fun.cr:60:3 in 'target_def_fun'
  from /crystal/src/compiler/crystal/codegen/call.cr:437:12 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/codegen.cr:2313:9 in 'codegen_assign'
  from /crystal/src/compiler/crystal/codegen/codegen.cr:997:7 in '??'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/semantic/bindings.cr:16:7 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/ast.cr:81:7 in 'codegen_fun'
  from /crystal/src/compiler/crystal/codegen/fun.cr:60:3 in 'target_def_fun'
  from /crystal/src/compiler/crystal/codegen/call.cr:437:12 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/codegen.cr:2330:7 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/enumerable.cr:463:7 in '??'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/ast.cr:81:7 in 'codegen_fun'
  from /crystal/src/compiler/crystal/codegen/fun.cr:60:3 in 'target_def_fun'
  from /crystal/src/compiler/crystal/codegen/call.cr:437:12 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/ast.cr:81:7 in 'codegen_fun'
  from /crystal/src/compiler/crystal/codegen/fun.cr:60:3 in 'target_def_fun'
  from /crystal/src/compiler/crystal/codegen/call.cr:437:12 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/ast.cr:81:7 in 'codegen_fun'
  from /crystal/src/compiler/crystal/codegen/fun.cr:60:3 in 'target_def_fun'
  from /crystal/src/compiler/crystal/codegen/call.cr:437:12 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/ast.cr:81:7 in 'codegen_fun'
  from /crystal/src/compiler/crystal/codegen/fun.cr:60:3 in 'target_def_fun'
  from /crystal/src/compiler/crystal/codegen/call.cr:437:12 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/semantic/bindings.cr:16:7 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/enumerable.cr:463:7 in '??'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/semantic/bindings.cr:16:7 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/enumerable.cr:463:7 in '??'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/codegen.cr:2313:9 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/codegen.cr:2313:9 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/codegen.cr:2313:9 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/ast.cr:81:7 in 'codegen_fun'
  from /crystal/src/compiler/crystal/codegen/fun.cr:60:3 in 'target_def_fun'
  from /crystal/src/compiler/crystal/codegen/call.cr:437:12 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/enumerable.cr:463:7 in '??'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/exception.cr:92:7 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/ast.cr:81:7 in 'codegen_fun'
  from /crystal/src/compiler/crystal/codegen/fun.cr:60:3 in 'target_def_fun'
  from /crystal/src/compiler/crystal/codegen/call.cr:437:12 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/semantic/bindings.cr:16:7 in 'codegen_cond'
  from /crystal/src/compiler/crystal/codegen/codegen.cr:908:7 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/enumerable.cr:463:7 in '??'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/ast.cr:81:7 in 'codegen_fun'
  from /crystal/src/compiler/crystal/codegen/fun.cr:60:3 in 'target_def_fun'
  from /crystal/src/compiler/crystal/codegen/call.cr:437:12 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/semantic/bindings.cr:16:7 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/enumerable.cr:463:7 in '??'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/ast.cr:81:7 in 'codegen_fun'
  from /crystal/src/compiler/crystal/codegen/fun.cr:60:3 in 'target_def_fun'
  from /crystal/src/compiler/crystal/codegen/call.cr:437:12 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/enumerable.cr:463:7 in '??'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/codegen.cr:729:15 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/codegen.cr:445:17 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/enumerable.cr:463:7 in '??'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/codegen.cr:445:17 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/enumerable.cr:463:7 in '??'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/codegen.cr:445:17 in 'visit'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/enumerable.cr:463:7 in '??'
  from /crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from /crystal/src/compiler/crystal/codegen/codegen.cr:2283:7 in 'codegen'
  from /crystal/src/compiler/crystal/compiler.cr:172:16 in 'compile:combine_rpath'
  from /crystal/src/compiler/crystal/command.cr:341:5 in 'run'
  from /crystal/src/compiler/crystal.cr:11:1 in '__crystal_main'
  from /crystal/src/crystal/main.cr:129:5 in 'main'
  from src/env/__libc_start_main.c:95:2 in 'libc_start_main_stage2'
Error: you've found a bug in the Crystal compiler. Please open an issue, including source code that will allow us to reproduce the bug: https://github.com/crystal-lang/crystal/issues
Error: Process completed with exit code 1.
```

Related: https://github.com/crystal-ameba/ameba/issues/397

--- a/src/io/buffered.cr
+++ b/src/io/buffered.cr
@@ -98,7 +98,7 @@ module IO::Buffered
   # peek data if the current buffer is empty:
   # otherwise no read is performed and whatever
   # is in the buffer is returned.
-  def peek : Bytes?
+  def peek : Bytes
     check_open
 
     if @in_buffer_rem.empty?

